implement M3, M4 & M5 spindle control commands (replaces PR #4939)

Bob-the-Kuhn opened the pull request:
This is a clone of PR #4939 with the conflicts resolved. 

There are no changes to the spindle specific sections.

All fixes/updates/improvements in PR #4939 are included. 

The following is from PR #4939. 

/////////////////////////////////////////////////////////////////////////////////////////////////////////

all code changes are in Marlin_main.cpp

none of the existing code was changed

section added to config_adv to enable/disable the new commands & to explain the hardware changes needed to properly connect to the spindle controller

There's lots of competition for the PWM pins so added a section in sanity.h to check that the selected pin doesn't conflict with other selections.

Bob-the-Kuhn committed:
replaces PR #4939

just resolving conflicts

no changes to spindle specific additions

Bob-the-Kuhn committed:
syntax correction

an attempt to fix the "All checks have failed" in the PR

Bob-the-Kuhn commented:
Just did a syntax correction to config_adv in an attempt to fix the "All checks have failed" issue in the PR.

thinkyhead commented on Marlin/SanityCheck.h:
```diff
@@ -1019,3 +1019,53 @@
+      #error "SPINDLE_SPEED pin in use by Z_STEP "
+    #elif defined(NUM_SERVOS) && (NUM_SERVOS > 0) &&  (SPINDLE_SPEED_PIN == 2 || SPINDLE_SPEED_PIN ==  3 || SPINDLE_SPEED_PIN == 5)
+      #error "Counter/Timer for SPINDLE_SPEED PWM pin is used by the servo system "
+    #elif PIN_EXISTS(EXTRUDER_0_AUTO_FAN) && SPINDLE_SPEED_PIN == EXTRUDER_0_AUTO_FAN_PIN
```
Names have changed to `E0_AUTO_FAN_PIN`, etc.

thinkyhead commented on Marlin/SanityCheck.h:
```diff
@@ -1019,3 +1019,53 @@
+#if ENABLED( SPINDLE_ENABLE)
+  #if !PIN_EXISTS(SPINDLE_ENABLE)
+    #error "SPINDLE_ENABLE requires SPINDLE_ENABLE_PIN."
+  #elif !defined(SPINDLE_ENABLE_INVERT) || !(SPINDLE_ENABLE_INVERT == true || SPINDLE_ENABLE_INVERT == false)
```
You can remove this, because it will (almost) always pass:

```
!(SPINDLE_ENABLE_INVERT == true || SPINDLE_ENABLE_INVERT == false)
```

That is to say, a zero value will evaluate to `false`, and since this setting is used as a `bool` any non-zero value will evaluate to `true`. I don't think we need to hand-hold here, since in the configuration it's already got a default value of `false`.

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -4478,6 +4481,133 @@ inline void gcode_G92() {
+    #endif
+  } 
+ 
+
```
One blank line is fine here.

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -4478,6 +4481,133 @@ inline void gcode_G92() {
+        delay_for_spindle_power_down();
+      }
+      else {
+        int16_t OCRxx_VAL = 2.55 * (spindle_speed - RPM_INTERCEPT)/RPM_SLOPE;      // convert RPM to PWM duty cycle
```
Minor formatting quibble. In most cases math and boolean operators should be preceded and followed by one space. See the coding standards document in the wiki.

Minor automatic variable naming quibble. They should be lowercase to contrast with macros (and _actual_ AVR interrupt addresses), so you can tell at a glance that they're variables. So here `ocr_val` is fine.

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -4478,6 +4481,133 @@ inline void gcode_G92() {
+
+  inline void delay_for_spindle_power_up() {
+    refresh_cmd_timeout();                                                       // wait for spindle to come up to speed
+    if (SPINDLE_POWER_UP_DELAY > 0)  while (PENDING(millis(), SPINDLE_POWER_UP_DELAY * 1000 + previous_cmd_ms)) idle();
```
Instead of this invariant code, add a sanity check that throws an error `#if SPINDLE_POWER_UP_DELAY < 1`. Then you only need:

``` cpp
while (PENDING(millis(), SPINDLE_POWER_UP_DELAY * 1000 + previous_cmd_ms)) idle();
```

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -4478,6 +4481,133 @@ inline void gcode_G92() {
+  
+  inline void delay_for_spindle_power_down() {
+    refresh_cmd_timeout();                                                       // wait for spindle to come up to speed
+    if (SPINDLE_POWER_DOWN_DELAY >= 0)  while (PENDING(millis(), SPINDLE_POWER_DOWN_DELAY * 1000 + previous_cmd_ms + 1)) idle(); 
```
Ditto with `SPINDLE_POWER_DOWN_DELAY`

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -4478,6 +4481,133 @@ inline void gcode_G92() {
+        if (temp >= 255)                OCRxx_VAL = 255;                                        //limit to max the Atmel PWM will support
+        if (spindle_speed <= RPM_MIN)   OCRxx_VAL = 2.55 * (RPM_MIN - RPM_INTERCEPT)/RPM_SLOPE; // minimum setting
+        if (spindle_speed >= RPM_MAX)   OCRxx_VAL = 2.55 * (RPM_MAX - RPM_INTERCEPT)/RPM_SLOPE; // limit to max RPM
+        if (SPINDLE_SPEED_INVERT==true) OCRxx_VAL = 255 - OCRxx_VAL ; 
```
You can leave off `==true` and there should be no extra space before the semicolon.

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -4478,6 +4481,133 @@ inline void gcode_G92() {
+        if (SPINDLE_SPEED_INVERT==true) OCRxx_VAL = 255 - OCRxx_VAL ; 
+        stepper.synchronize();   // wait until previous movement commands (G0, G1, G2 & G3) have completed before turning spindle on
+        WRITE(SPINDLE_ENABLE_PIN, SPINDLE_ENABLE_INVERT);  // turn spindle on (active low)
+        analogWrite(SPINDLE_SPEED_PIN, OCRxx_VAL & 0xff);  //only write lowest byte 
```
Uppercase hexadecimal values are preferred in Marlin. `0xFF`.

thinkyhead commented on Marlin/SanityCheck.h:
```diff
@@ -1019,3 +1019,53 @@
+      #error "SPINDLE_SPEED PWM pin is used by EXTRUDER_3_AUTO_FAN_PIN"
+    #elif PIN_EXISTS(FAN) && SPINDLE_SPEED_PIN == FAN_PIN 
+      #error "SPINDLE_SPEED PWM pin is used FAN_PIN "
+    #elif PIN_EXISTS(FAN0) && SPINDLE_SPEED_PIN == FAN0_PIN 
```
There is no `FAN0_PIN` so this can be removed.

thinkyhead commented on Marlin/SanityCheck.h:
```diff
@@ -1019,3 +1019,53 @@
+      #error "SPINDLE_SPEED PWM pin is used by MOTOR_CURRENT_PWM_E"
+    #endif  
+  #endif
+#endif   //SPINDLE_ENABLE
```
One space before and after `//`.
And add a carriage return.

thinkyhead commented on Marlin/SanityCheck.h:
```diff
@@ -1019,3 +1019,53 @@
+    #error "SPINDLE_ENABLE requires SPINDLE_ENABLE_PIN."
+  #elif !defined(SPINDLE_ENABLE_INVERT) || !(SPINDLE_ENABLE_INVERT == true || SPINDLE_ENABLE_INVERT == false)
+    #error "SPINDLE_ENABLE_INVERT missing or incorrectly defined"
+  #elif ENABLED (SPINDLE_SPEED) && PIN_EXISTS(SPINDLE_SPEED)
```
No space between `ENABLED` and the opening parenthesis.

thinkyhead commented on Marlin/SanityCheck.h:
```diff
@@ -1019,3 +1019,53 @@
+      #error "spindle speed equation constant(s) missing"
+    #elif SPINDLE_SPEED_PIN == 4 || SPINDLE_SPEED_PIN == 11 || SPINDLE_SPEED_PIN == 12 || SPINDLE_SPEED_PIN == 13
+      #error "Counter/Timer for SPINDLE_SPEED PWM pin is used by a system interrupt "
+    #elif PIN_EXISTS(X_MAX) &&  X_MAX_PIN == SPINDLE_SPEED_PIN && defined(USE_XMAX_PLUG)
```
We've already established that `SPINDLE_SPEED_PIN` is something other than -1. We also know that `X_MAX_PIN` exists. We also know that if `USE_XMAX_PLUG` is set that `X_MAX_PIN` cannot be -1. So you only need `X_MAX_PIN == SPINDLE_SPEED_PIN` here.

Only one space after `&&`. I will not merge a PR that has even a single space out of place, tabs instead of spaces, or trailing whitespace.

thinkyhead commented on Marlin/SanityCheck.h:
```diff
@@ -1019,3 +1019,53 @@
+    #elif PIN_EXISTS(X_MAX) &&  X_MAX_PIN == SPINDLE_SPEED_PIN && defined(USE_XMAX_PLUG)
+      #error "SPINDLE_SPEED pin is in use by X_MAX endstop"
+    #elif PIN_EXISTS(X_MIN) &&  X_MIN_PIN == SPINDLE_SPEED_PIN && defined(USE_XMIN_PLUG)
+      #error "SPINDLE_SPEED pin is in use by X_MIN endstop "
```
Watch those extra spaces. If formatting isn't perfect, the PR will not be merged.

thinkyhead commented on Marlin/SanityCheck.h:
```diff
@@ -1019,3 +1019,53 @@
+    #elif !defined(SPINDLE_SPEED_INVERT) || !(SPINDLE_SPEED_INVERT == true || SPINDLE_SPEED_INVERT == false)
+      #error "SPINDLE_SPEED_INVERT missing or incorrectly defined"
+    #elif !defined(RPM_SLOPE) || !defined(RPM_INTERCEPT) || !defined(RPM_MIN) || !defined(RPM_MAX)
+      #error "spindle speed equation constant(s) missing"
```
Capitalize the first letter in error messages. End error messages with a period.

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -9715,6 +9860,13 @@ void setup() {
+  #if ENABLED(SPINDLE_ENABLE)
+    OUT_WRITE(SPINDLE_ENABLE_PIN, !SPINDLE_ENABLE_INVERT);  //init spindle to off;
+    #if ENABLED(SPINDLE_SPEED)
+      pinMode(SPINDLE_SPEED_PIN, OUTPUT);
```
Use fastio `SET_OUTPUT(SPINDLE_SPEED_PIN)` here.

thinkyhead reviewed (changes requested):
Various changes needed. Especially remove all the weird extra spaces strewn throughout.

Bob-the-Kuhn committed:
changes from review and synched with latest RCBugFix

1) incorporated all the changes requested by the review

2) added in all the changes between the original fork and the latest
RCBugFix

Bob-the-Kuhn committed:
brought config_adv up todate with RCBugfix

Bob-the-Kuhn commented:
replaced by PR #5145

thinkyhead commented on Marlin/Configuration_adv.h:
```diff
@@ -829,6 +829,106 @@
+ *  You'll need to select a 3.3V - 5V signal for the ON/OFF function and optionally
+ *  choose a 0-5V PWM pin for the speed control.
+ *
+ *  uncomment the following to enable the ON/OFF function
```
We don't need "uncomment to enable" on every setting. We assume users get that already.

thinkyhead commented on Marlin/Configuration_adv.h:
```diff
@@ -829,6 +829,106 @@
+ *  uncomment the following to enable the ON/OFF function
+ */
+ 
+//#define SPINDLE_ENABLE
```
All settings that depend on `SPINDLE_ENABLE` should be inside a block that starts with `#if ENABLED(SPINDLE_ENABLE)`. This allows a putative configuration utility to glean which settings depend on others, and allows users to see this also at a glance.

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -860,11 +864,17 @@ void setup_homepin(void) {
-    OUT_WRITE(PHOTOGRAPH_PIN, LOW);
-  #endif
-}
+#if HAS_CASE_LIGHT
```
If `M355` stuff is showing up then this PR needs to be rebased and commits squashed or skipped. `M355` has already been merged.
